### PR TITLE
Adding shadow for `readParcelCreator` to overcome JDK7 field access issue

### PR DESCRIPTION
**Pull request by menny**
This is related to https://github.com/robolectric/robolectric/issues/2398

### CODE REVIEW

**Comment by menny** [comment]
Adding `java.lang.reflect.*` or just `java.lang.reflect.Field` failed miserably with 10375 unit-test errors :scream: 
```
testExecSQL[19](org.robolectric.shadows.SQLiteDatabaseTest)  Time elapsed: 0 sec  <<< ERROR!
java.util.ServiceConfigurationError: org.robolectric.internal.ShadowProvider: Provider org.robolectric.Shadows not a subtype
	at java.util.ServiceLoader.fail(ServiceLoader.java:239)
	at java.util.ServiceLoader.access$300(ServiceLoader.java:185)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:376)
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
	at org.robolectric.internal.bytecode.InstrumentationConfiguration$Builder.build(InstrumentationConfiguration.java:149)
	at org.robolectric.RobolectricTestRunner.createClassLoaderConfig(RobolectricTestRunner.java:132)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:185)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:54)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:193)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:52)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:191)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:42)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:184)
	at org.robolectric.RobolectricTestRunner$1.evaluate(RobolectricTestRunner.java:152)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:236)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:24)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:193)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:52)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:191)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:42)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:184)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:236)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray2(ReflectionUtils.java:208)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:158)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:86)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:95)
```

**Comment by jongerrish** [comment]
Can you post the original stack trace? What you posted is just the
subsequent failures...it's really annoying actually we should fix
that...but basically is the failure on subsequent tests when related to a
previous failure setting up the environment.

**Comment by menny** [comment]
@jongerrish I've made the code more generic. Still don't think it is the correct solution.

**Comment by menny** [comment]
@jongerrish @xian @erd can you take a look at this? Your thoughts?

**Comment by menny** [comment]
@xian can you take a look at this PR?

**Comment by menny** [comment]
@xian not really sure what's going on with circle-ci. It doesn't start the build, and I can not find the PR (or branch) at https://circleci.com/gh/robolectric/robolectric

**Comment by xian** [comment]
@menny I've changed circleci to build forked PR's, can you merge master and try again?

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowParcel.java [hunk 1]
```diff
@@ -23,13 +26,48 @@
 import org.robolectric.annotation.RealObject;
 import org.robolectric.util.ReflectionHelpers;
 
+/**
```
Comment (xian) [inline shadows/framework/src/main/java/org/robolectric/shadows/ShadowParcel.java]: No javadoc needed here (we only want javadoc describing differences in the shadow's overall behavior).

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowParcel.java [hunk 2]
```diff
@@ -23,13 +26,48 @@
...
+ */
 @Implements(Parcel.class)
 @SuppressWarnings("unchecked")
 public class ShadowParcel {
   @RealObject private Parcel realObject;
   private static final Map<Long, ByteBuffer> NATIVE_PTR_TO_PARCEL = new LinkedHashMap<>();
   private static long nextNativePtr = 1; // this needs to start above 0, which is a magic number to Parcel
 
+  @Implementation(maxSdk = JELLY_BEAN_MR2)
+  public <T extends Parcelable> T readParcelable(ClassLoader loader) throws IllegalAccessException, NoSuchFieldException, ClassNotFoundException {
```
Comment (xian) [inline shadows/framework/src/main/java/org/robolectric/shadows/ShadowParcel.java]: This looks like it duplicates behavior from Jelly Bean's `readParcelable()`, no? Is it necessary, or is implementing `readParcelableCreator()` enough?

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowParcel.java [hunk 3]
```diff
@@ -23,13 +26,48 @@
...
+
+    if (creator instanceof Parcelable.ClassLoaderCreator<?>) {
+      Parcelable.ClassLoaderCreator<?> classLoaderCreator = (Parcelable.ClassLoaderCreator<?>) creator;
+      return (T) classLoaderCreator.createFromParcel(realObject, loader);
+    }
+    return (T) creator.createFromParcel(realObject);
+  }
+
+  @Implementation @HiddenApi
+  public final Parcelable.Creator<?> readParcelableCreator(ClassLoader loader) throws ClassNotFoundException, NoSuchFieldException, IllegalAccessException {
```
Comment (xian) [inline shadows/framework/src/main/java/org/robolectric/shadows/ShadowParcel.java]: Remove `final`?

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowParcel.java (hunk 1 above)
Comment (menny) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowParcel.java]: done

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowParcel.java (hunk 2 above)
Comment (menny) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowParcel.java]: I was wrong, it should be `JELLY_BEAN_MR1`.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowParcel.java (hunk 3 above)
Comment (menny) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowParcel.java]: done

**Comment by menny** [comment]
@xian thanks.

**Pull request merged**